File: frontend/src/types/borrow.ts
import type { BaseQuery } from "./baseQuery";
import type { Item } from "./item";
import type { User } from "./user";

export type BorrowStatus =
  | "pending"
  | "rejected"
  | "approved"
  | "borrowed"
  | "returned"
  | "canceled";

export interface Borrow {
  id: number;
  user: User | null;
  reviewedUser: User | null;
  item: Item | null;
  purpose: string;
  quantity: number;
  officerNote: string | null;
  borrowDate: string;
  returnDate: string;
  status: BorrowStatus;
  reviewAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface BorrowQuery extends BaseQuery {
  status?: BorrowStatus;
  startDate?: string;
  endDate?: string;
}

export interface BorrowRequest {
  itemId: number;
  quantity: number;
  purpose: string;
  borrowDate: string | null;
  returnDate: string | null;
}

export interface BorrowForUserRequest extends BorrowRequest {
  userId: number;
}

export interface BorrowError {
  userId: number;
  itemId: number;
  quantity: number;
  purpose: string;
  borrowDate: string;
  returnDate: string;
}

export interface BorrowApprovalRequest {
  officerNote: string | null;
}

export interface BorrowCard {
  totalPending: number;
  totalApproved: number;
  totalRejected: number;
}

File: backend/internal/model/borrow.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type BorrowStatus string

const (
	BorrowStatusPending  BorrowStatus = "pending"
	BorrowStatusRejected BorrowStatus = "rejected"
	BorrowStatusApproved BorrowStatus = "approved"
	BorrowStatusBorrowed BorrowStatus = "borrowed"
	BorrowStatusReturned BorrowStatus = "returned"
	BorrowStatusCanceled BorrowStatus = "canceled"
)

type Borrow struct {
	ID             uint         `gorm:"primaryKey"`
	UserID         uint         `gorm:"not null;index"`
	User           User         `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE"`
	ReviewedUserID *uint        `gorm:"index"`
	ReviewedUser   *User        `gorm:"foreignKey:ReviewedUserID;constraint:OnDelete:SET NULL"`
	ItemID         uint         `gorm:"not null;index"`
	Item           Item         `gorm:"constraint:OnDelete:CASCADE"`
	purpose        string       `gorm:"not null"`
	Quantity       int          `gorm:"not null"`
	OfficerNote    *string      `gorm:"type:text"`
	BorrowDate     time.Time    `gorm:"type:date;not null"`
	ReturnDate     time.Time    `gorm:"type:date;not null"`
	Status         BorrowStatus `gorm:"type:enum('pending', 'rejected', 'approved', 'borrowed', 'returned', 'canceled');default:pending;not null"`
	ReviewAt       *time.Time   `gorm:"type:date"`
	CreatedAt      time.Time
	UpdatedAt      time.Time
	DeletedAt      gorm.DeletedAt `gorm:"index"`
}
